### ISSUE TITLE
No autocompletions for Parent class when using ES2015 mixins (function returning the Class)

### ISSUE THREAD

**Issue by kasperpeulen** [Microsoft/TypeScript#18732]
Consider this code:
```javascript
class Parent {
  parentMethod() {  }
}

const Mixin = Base => class Mixin extends Base {
  mixinMethod() {  }
}

class Child extends Mixin(Parent) {
  childMethod() {
    this.
  }
}
```

For the Child class, all three method are available, but typescript doesn't detect the `parentMethod` to be available.

![image](https://user-images.githubusercontent.com/1035299/30789089-4e58907c-a1a3-11e7-9b75-c25f5cf11a70.png)

**Comment by jwbay** [comment]
There may be a better way to do this, but the main problem is that Base is of type `any` for your mixin. If you give it a generic type so it can flow through, parent methods will show up.

```ts
type Constructable = new (...args) => any;

class Parent {
  parentMethod() {  }
}

const Mixin = <T extends Constructable>(Base: T) => class Mixin extends Base {
  mixinMethod() {  }
}

class Child extends Mixin(Parent) {
  childMethod() {
    this. // all three show up here in playground
  }
}
```

**Comment by kasperpeulen** [comment]
Nice! I tried to implement this using jsdoc syntax, but without luck, any ideas?

```
/** @typedef {new (...args) => any} Constructable */

class Parent {
  parentMethod() {  }
}

/**
 * @template T extends Constructable
 * @param {T} Base 
 */
const Mixin = (Base) => class Mixin extends Base {
  mixinMethod() {  }
}

class Child extends Mixin(Parent) {
  childMethod() {
    this. // still only two autocompletions
  }
}
```

**Comment by DanielRosenwasser** [comment]
I think the problem here is that TypeScript doesn't have a way of expressing constraints on type parameters in JSDoc.

In other words:

```js
/**
 * @template T extends Constructable
 */
```

doesn't actually add a constraint.

@dimvar I see that you mentioned some work on bounded generics [here](https://github.com/google/closure-compiler/issues/576#issuecomment-211489044) for Closure. Any word on that?

**Comment by dimvar** [comment]
No progress on that yet.

**Comment by akdor1154** [comment]
Is there anywhere that "types I can describe in TS but not JSDoc" are listed?

**Comment by ma2ciek** [comment]
Hi!

Is there some progress on that field (I mean allowing constraints in the JSDoc `@template`)? It'd be really helpful as it could allow type checking JS code based on mixins.

In many scenarios, the previous approach (https://www.typescriptlang.org/docs/handbook/mixins.html) is not an option as it requires a massive code duplication (e.g. creating 15 fake properties and methods to satisfy interfaces).

**Comment by RyanCavanaugh** [comment]
OP works in latest on Playground

**Comment by Rampage0131** [comment]
@RyanCavanaugh 
Please explain more details
